guvnor-ala/guvnor-ala-build-maven/src/main/java/org/guvnor/ala/build/maven/util/RepositoryVisitor.java

```diff
@@ -13,109 +13,221 @@
+        } catch (IOException ex) {
+            Logger.getLogger(RepositoryVisitor.class.getName()).log(Level.SEVERE, null, ex);
+        }
     }
 
-    public File getGuvnorTempFolder() {
-        return new File( System.getProperty( "java.io.tmpdir" ) + File.separatorChar + "guvnor" );
-    }
+    private void loadIndex(String path) {
+        Gson gson = new Gson();
+        try {
+            JsonReader reader = new JsonReader(new FileReader(path + "/index.json"));
+            Type type = new TypeToken<Map<String, String>>() {
+            }.getType();
+            oldIdentityHash = gson.fromJson(reader, type);
+        } catch (FileNotFoundException ex) {
+            //The Index doesn't exist yet, it will be generated after the Repository visitor is used for the first time.
+        }
 
-    public File getTargetFolder() {
-        return new File( buildRoot + File.separatorChar + project.getRootPath().toAbsolutePath()
-                + File.separatorChar + project.getPath().toAbsolutePath().toString() + File.separatorChar + "target" );
     }
 
-    private void visitPaths( final DirectoryStream<Path> directoryStream ) throws IOException {
-        for ( final org.uberfire.java.nio.file.Path path : directoryStream ) {
-            if ( Files.isDirectory( path ) ) {
-                makeTempDirectory( path );
-                visitPaths( Files.newDirectoryStream( path ) );
+    private void visitPaths(final File parent,
+            final DirectoryStream<Path> directoryStream) throws IOException, NoSuchAlgorithmException {
+        for (final org.uberfire.java.nio.file.Path path : directoryStream) {
+            if (Files.isDirectory(path)) {
+                final File newParent = makeTempDirectory(parent, path.getFileName().toString());
+                visitPaths(newParent, Files.newDirectoryStream(path));
             } else {
-                makeTempFile( path );
+                makeTempFile(parent, path);
             }
         }
     }
 
-    private File makeTempDirectory( Path path ) {
-        return makeTempDirectory( getFilePath( path ) );
-    }
-
-    private File makeTempDirectory( String filePath ) {
-        File tempDirectory = new File( filePath );
-        if ( !tempDirectory.isFile() ) {
+    private File makeTempDirectory(final File parent,
+            final String filePath) {
+        File tempDirectory = new File(parent, filePath);
+        if (tempDirectory.exists()) {
+            return tempDirectory;
+        }
+        if (!tempDirectory.isFile()) {
             tempDirectory.mkdir();
         }
         return tempDirectory;
     }
 
-    private void makeTempRootDirectory() {
-        File tempDirectory = new File( buildRoot );
-        tempDirectory.mkdirs();
+    private File makeTempRootDirectory(final String tempRoot,
+            final boolean recreateTempDir) {
+        final File tempRootDir = new File(tempRoot);
+        if (tempRootDir.exists()) {
+            if (recreateTempDir) {
```

Comment:
Okay. So really this method is about cleaning a reused directory, right? If that's the case I would suggest renaming the method and parameter to use the word "clean" instead of "make" and "recreate".